### PR TITLE
HPCC-13349 New ECLWatch reports expired HTPASSWD

### PR DESCRIPTION
**Pull request by RussWhitehead**
Currently any authentication method other than LDAP returns the "Expired"
flag for the password. This PR changes that to return the "Never Expires"
flag

Signed-off-by: William Whitehead william.whitehead@lexisnexis.com

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@miguelvazq Please review

**Comment by miguelvazq** [comment]
Looks good.

**Comment by RussWhitehead** [comment]
@ghalliday @richardkchapman Please merge and close

File: system/security/shared/SecureUser.hpp [hunk 1]
```diff
@@ -200,7 +200,7 @@ class CSecureUser : public CInterface,
 
     virtual CDateTime & getPasswordExpiration(CDateTime& expirationDate){ return expirationDate; }
     virtual bool setPasswordExpiration(CDateTime& expirationDate) { return true; }
-    virtual int getPasswordDaysRemaining() {return -1;}
+    virtual int getPasswordDaysRemaining() {return -2;}//never expires
```
Comment (ghalliday) [inline system/security/shared/SecureUser.hpp]: There should be some constants in seclib.hpp to make these magic values clearer

e.g.,

const static int PasswordExpired = -1;
const static int PasswordNeverExpires = -2;

Could possibly do as a separate commit though.  If so please open a jira and I'll merge this.

Comment (ghalliday) [inline system/security/shared/SecureUser.hpp]: The constants should be used in espprotocol and the other places that call this function...

**Comment by RussWhitehead** [comment]
@ghalliday Please review

File: system/security/shared/seclib.hpp [hunk 2]
```diff
@@ -116,6 +116,8 @@ enum SecUserStatus
 };
 
 
+const static int scPasswordExpired = -1;
```
Comment (ghalliday) [inline system/security/shared/seclib.hpp]: Trivial: I assume the sc is for static const.  We don't really use that naming style inside the main platform - although I know esp does.

File: system/security/shared/SecureUser.hpp [hunk 3]
```diff
@@ -200,7 +200,7 @@ class CSecureUser : public CInterface,
 
     virtual CDateTime & getPasswordExpiration(CDateTime& expirationDate){ return expirationDate; }
     virtual bool setPasswordExpiration(CDateTime& expirationDate) { return true; }
-    virtual int getPasswordDaysRemaining() {return -1;}
+    virtual int getPasswordDaysRemaining() {return scPasswordNeverExpires;}//never expires
```
Comment (ghalliday) [inline system/security/shared/SecureUser.hpp]: trivial: comment is now redundant

File: system/security/LdapSecurity/ldapsecurity.ipp [hunk 4]
```diff
@@ -110,13 +110,13 @@ public:
    virtual int getPasswordDaysRemaining()
    {
        if (m_passwordExpiration.isNull())
-           return -2;//-2 if never expires
+           return scPasswordNeverExpires;//-2 if never expires
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapsecurity.ipp]: trivial: comment is now redundant

File: system/security/LdapSecurity/ldapsecurity.ipp [hunk 5]
```diff
@@ -110,13 +110,13 @@ public:
...
        CDateTime expiry(m_passwordExpiration);
        CDateTime now;
        now.setNow();
        now.adjustTime(now.queryUtcToLocalDelta());
        if (expiry <= now)
-           return -1;//-1 if already expired
+           return scPasswordExpired;//-1 if already expired
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapsecurity.ipp]: trivial: comment is now redundant

File: esp/platform/espprotocol.cpp [hunk 6]
```diff
@@ -120,8 +120,8 @@ const StringBuffer &CEspApplicationPort::getAppFrameHtml(time_t &modified, const
         {
             passwordDaysRemaining = user->getPasswordDaysRemaining();//-1 if expired, -2 if never expires
```
Comment (ghalliday) [inline esp/platform/espprotocol.cpp]: trivial: comment not really needed any more

File: esp/platform/espprotocol.cpp [hunk 7]
```diff
@@ -112,7 +112,7 @@ const StringBuffer &CEspApplicationPort::getAppFrameHtml(time_t &modified, const
 
     if (needRefresh || embedded_url || !appFrameHtml.length())
     {
-        int passwordDaysRemaining = -1;//-1 means dont display change password screen
+        int passwordDaysRemaining = scPasswordExpired;//-1 means dont display change password screen
```
Comment (ghalliday) [inline esp/platform/espprotocol.cpp]: minor: comment is out of date.  It seems slightly counter intuitive that expired means that, but I don't know the context.

**Comment by ghalliday** [comment]
@RussWhitehead I merged as is.  If you want to clean up the comments then open another pull request (with same issue is fine)

Also, did you get my emails about ldap questions?

File: system/security/shared/seclib.hpp (hunk 2 above)
Comment (RussWhitehead) [inline system/security/shared/seclib.hpp]: Actually it was for "SecCredentials" but I see your point

**Pull request merged**